hail/src/main/scala/is/hail/io/LoadMatrix.scala

```diff
@@ -337,64 +140,334 @@ object LoadMatrix {
+      fileByPartition,
+      firstPartitions,
+      hasHeader)
+    val rdd = ContextRDD.weaken[RVDContext](lines.filter(l => l.value.nonEmpty))
+      .cmapPartitionsWithIndex(compiledLineParser)
+    val rvd = if (tr.dropRows)
+      RVD.empty(sc, requestedType.canonicalRVDType)
+    else
+      RVD.unkeyed(requestedType.rowType.physicalType, rdd)
+    val globalValue = makeGlobalValue(ctx, requestedType, colIDs.map(Row(_)))
+    TableValue(tr.typ, globalValue, rvd)
+  }
+}
 
-    if (!keyFields.forall(rowType.fieldNames.contains))
-      fatal(
-        s"""Some row keys not found in row schema:
-           |    '${ keyFields.filter(!rowType.fieldNames.contains(_)).mkString("'\n    '") }'
-         """.stripMargin
-      )
-
-    val matrixType = MatrixType(
-      TStruct.empty(),
-      colType = TStruct("col_id" -> (if (noHeader) TInt32() else TString())),
-      colKey = Array("col_id"),
-      rowType = rowType,
-      rowKey = rowKey.toFastIndexedSeq,
-      entryType = cellType)
-
-    val tt = matrixType.canonicalTableType
-    val rvdType = tt.canonicalRVDType
+class MatrixParseError(
+  val msg: String,
+  val filename: String,
+  val line: Long,
+  val posStart: Int,
+  val posEnd: Int
+) extends RuntimeException(s"${filename}:${posStart}-${posEnd}, ${msg}")
+
+class CompiledLineParser(
+  requestedTableType: TableType,
+  nCols: Int,
+  missingValue: String,
+  separator: Char,
+  partitionCounts: Array[Long],
+  fileByPartition: Array[String],
+  firstPartitions: Array[Int],
+  hasHeader: Boolean
+) extends ((Int, RVDContext, Iterator[WithContext[String]]) => Iterator[RegionValue]) with Serializable {
+  assert(!missingValue.contains(separator))
+  @transient private[this] val requestedRowType = requestedTableType.canonicalPType
+  @transient private[this] val entriesType = requestedRowType
+    .selfField(MatrixType.entriesIdentifier)
+    .map(f => f.typ.asInstanceOf[PArray])
+  @transient private[this] val rowFieldsType = requestedRowType
+    .dropFields(Set(MatrixType.entriesIdentifier))
+  @transient private[this] val fb = new Function4Builder[Region, String, Long, String, Long]
+  @transient private[this] val mb = fb.apply_method
+  @transient private[this] val region = fb.arg1
+  @transient private[this] val _filename = fb.arg2
+  @transient private[this] val _lineNumber = fb.arg3
+  @transient private[this] val _line = fb.arg4
+  @transient private[this] val filename = mb.newField[String]("filename")
+  @transient private[this] val lineNumber = mb.newField[Long]("lineNumber")
+  @transient private[this] val line = mb.newField[String]("line")
+  @transient private[this] val pos = mb.newField[Int]("pos")
+  @transient private[this] val srvb = new StagedRegionValueBuilder(mb, requestedRowType)
+  fb.addInitInstructions(Code(
+    pos := 0,
+    filename := Code._null,
+    lineNumber := 0,
+    line := Code._null,
+    srvb.init()))
+
+  @transient private[this] val parseStringMb = fb.newMethod[Region, String]
```

Comment:
you can add string names to all these methods to make profiling a lot easier! I'm going to do that with as much of the codebase as possible soon.

Reply:
I went down this path but it changed several files and looked like EmitFunctionBuilder needed a refactor to reduce some duplicate code so I backed it all out after I finished profiling. Happy to add back in if another PR adds the functionality.

Reply:
can't you just do `fb.newMethod[Region, String]("parseStringMB")`?